### PR TITLE
Screens: Legacy settings screen

### PR DESCRIPTION
**Pull request by kristofbolyai**
TODO:
- [x] Categories
- [x] Search
- [x] Actual settings screen

### CODE REVIEW

**Comment by kristofbolyai** [comment]
2022.10.02:
image

**Comment by kristofbolyai** [comment]
This will be likely the cleanest UI in terms of code. I have figured out every nit in making GUIs like this, so I can avoid mess even before I create it.

**Comment by kristofbolyai** [comment]
image
image
image
image

**Comment by Incompleteusern** [comment]
LGTM, I will review when I have the time

File: common/src/main/java/com/wynntils/gui/render/RenderUtils.java [hunk 1]
```diff
@@ -217,6 +217,36 @@ public static void drawRect(
...
+                0,
+                hovered ? texture.height() / 2f : 0,
+                texture.width(),
+                texture.height() / 2f,
+                texture.width(),
+                texture.height());
+    }
+
+    public static void drawTexturedRect(PoseStack poseStack, Texture texture, float x, float y) {
+        RenderUtils.drawTexturedRect(
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/render/RenderUtils.java]: The prefix here and above is unnecessary

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    // region render
+
+    @Override
+    public void render(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        float backgroundRenderX = getTranslationX();
+        float backgroundRenderY = getTranslationY();
+
+        poseStack.pushPose();
+        poseStack.translate(backgroundRenderX, backgroundRenderY, 0);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:116]: Perhaps this could be a separate method called `translateBackground`?

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        mouseX -= getTranslationX();
+        mouseY -= getTranslationY();
+
+        for (GuiEventListener child : children()) {
+            if (child.isMouseOver(mouseX, mouseY)) {
+                child.mouseClicked(mouseX, mouseY, button);
+            }
+        }
+
+        for (int i = scrollOffset; i < Math.min(features.size(), scrollOffset + FEATURES_PER_PAGE); i++) {
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java]: `Math.min(features.size(), scrollOffset + FEATURES_PER_PAGE); ` could be extracted to a method

**Comment by Benj2005** [comment]
There seems to be a lot of settings missing from the minimap compared to the old version. Is this intentional?
**Edit: Fixed by 9fa4985d2fb07d6af8b87073827d846a866fef4f**
WynntilsUI

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        scrollOffset =
+                MathUtils.clamp((int) (scrollOffset - delta), 0, Math.max(0, features.size() - FEATURES_PER_PAGE));
+    }
+
+    // endregion
+
+    @Override
+    public void onClose() {
+        ConfigManager.loadConfigFile();
+        ConfigManager.loadConfigOptions(ConfigManager.getConfigHolders(), true);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:328]: This should only be called when the "close" button is called, and not when the config is saved.

**Review by Incompleteusern — COMMENTED** [review]
More comments

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public void renderButton(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        FontRenderer.getInstance()
+                .renderAlignedTextInBox(
+                        poseStack,
+                        I18n.get(featureCategory.toString()),
+                        this.x,
+                        this.x + this.width,
+                        this.y,
+                        0,
+                        CommonColors.GREEN,
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: I think it shouldn't be green but a more neutral color. Also, it might be nice to make the button special further, by addding a box, shadowing it, or bolding it

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 6]
```diff
@@ -113,12 +145,24 @@ private void renderButtons(PoseStack poseStack, int mouseX, int mouseY, float pa
...
+        if (configListScrollButton != null) {
+            configListScrollButton.renderButton(poseStack, mouseX, mouseY, partialTick);
+        }
+
+        // Reverse iteration for optimal rendering
+        for (int i = Math.min(configs.size(), configScrollOffset + CONFIGS_PER_PAGE) - 1;
+                i >= configScrollOffset;
+                i--) {
+            AbstractButton configButton = configs.get(i);
+            configButton.render(poseStack, mouseX, mouseY, partialTick);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:186]: How is this more optimal?

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 7]
```diff
@@ -30,16 +37,21 @@
...
+    private final List<AbstractButton> configs = new ArrayList<>();
 
     private TextInputBoxWidget focusedTextInput;
     private final SearchWidget searchWidget;
     private ScrollButton featureListScrollButton;
+    private ScrollButton configListScrollButton;
     private Feature selected = null;
-    private int scrollOffset = 0;
+    private int featureScrollOffset = 0;
+    private int configScrollOffset = 0;
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:62]: A decent bit of code duplication, though likely can not be fixed

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java [hunk 8]
```diff
@@ -261,8 +266,11 @@ private void scrollFeatureList(double delta) {
     }
 
     private void scrollConfigList(double delta) {
-        configScrollOffset =
-                MathUtils.clamp((int) (configScrollOffset - delta), 0, Math.max(0, configs.size() - CONFIGS_PER_PAGE));
+        int roundedUpPageNeed = configs.size() / CONFIGS_PER_PAGE + (configs.size() % CONFIGS_PER_PAGE == 0 ? 0 : 1);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:296]: This is static, cache it

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/ConfigButton.java [hunk 9]
```diff
@@ -21,21 +28,29 @@
...
         this.resetButton = new GeneralSettingsButton(
                 this.x + 3,
                 this.y + 13,
                 35,
                 12,
                 new TranslatableComponent("screens.wynntils.settingsScreen.reset.name"),
-                configHolder::reset,
+                () -> {
+                    configHolder.reset();
+                    this.configOptionElement = getWidgetFromConfigHolder(configHolder);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/gui/screens/settings/widgets/ConfigButton.java:50]: Why is this necessary?

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 6 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:186]: Backwards rendering fixes render overlapping of tooltpis

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/ConfigButton.java (hunk 9 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/ConfigButton.java:50]: Reload text inputs after reset

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java (hunk 5 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: Idk design stuff... If you have a concrete suggestion lmk

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 3 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java]: Sometimes it is more readable to have it duplicated 1-2 times

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java]: imo

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:328]: We do this with overlays too. We save and then reload the current config state. Perhaps this is not needed though. Idk why I did this the first time.

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 6 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:186]: Change comment to reflect that?

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java (hunk 5 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: Try coloring it yellow maybe?

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 3 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java]: That's acceptable

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java (hunk 5 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: yellow is hover color

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: I don't known then, as green is enabled color and confusing

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 8 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:296]: Same, I think it might create mess if cached.

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 7 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:62]: Yeah...

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 2 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:116]: No need for a method, we only do this here

File: common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java (hunk 5 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/widgets/CategoryButton.java]: Changed it to cyan.

File: common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/WynntilsBookSettingsScreen.java:328]: I though about this and I like the way it is now. We re-load every config option, to update their states.

**Review by Incompleteusern — DISMISSED** [review]
Im satisfied besides the selectedFeature/selectedOverlay, but that has also been agreed upon.

**Review by Incompleteusern — APPROVED** [review]

File: common/src/main/java/com/wynntils/gui/screens/settings/elements/CustomColorConfigOptionElement.java [hunk 10]
```diff
@@ -25,7 +29,7 @@ public void renderConfigAppropriateButton(
 
         CustomColor value = (CustomColor) configHolder.getValue();
 
-        RenderUtils.drawRect(poseStack, value, 105, 0, 0, renderHeight, renderHeight);
+        RenderUtils.drawRect(poseStack, value, 85, 0, 0, renderHeight, renderHeight);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/gui/screens/settings/elements/CustomColorConfigOptionElement.java:32]: Qodana says:

 Suspicious variable/parameter name combination

'renderHeight' should probably not be passed as parameter 'width'

What do you have to say to your defense? :)

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/gui/screens/settings/elements/CustomColorConfigOptionElement.java:32]: Square.

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 11]
```diff
@@ -436,10 +436,10 @@
   "feature.wynntils.wynncraftButton.name": "Wynncraft Button",
   "feature.wynntils.wynncraftPauseScreen.classSelectionButton.name": "Class Selection",
   "feature.wynntils.wynncraftPauseScreen.hubButton.name": "Back to Hub",
-  "feature.wynntils.wynncraftPauseScreen.name": "Wynncraft Custom Pause Screen",
+  "feature.wynntils.wynncraftPauseScreen.name": "Wynntils Pause Screen",
   "feature.wynntils.wynncraftPauseScreen.wynntilsMenuButton.name": "Wynntils Menu",
-  "feature.wynntils.wynntilsPauseScreen.name": "Wynntils Custom Pause Screen",
-  "feature.wynntils.wynntilsQuestBook.name": "Wynncraft Custom Quest Book",
+  "feature.wynntils.wynntilsPauseScreen.name": "Wynntils Pause Screen",
```
Comment (magicus) [inline common/src/main/resources/assets/wynntils/lang/en_us.json:441]: This is the same key twice..?

Comment (kristofbolyai) [reply common/src/main/resources/assets/wynntils/lang/en_us.json:441]: wynncraft vs wynntils. I plan to clean up i18n file as there is a few unused keys. Perhaps make a script even.

Comment (magicus) [reply common/src/main/resources/assets/wynntils/lang/en_us.json:441]: Aha.

Comment (magicus) [reply common/src/main/resources/assets/wynntils/lang/en_us.json:441]: And yes, good idea!

File: common/src/main/java/com/wynntils/gui/screens/settings/elements/CustomColorConfigOptionElement.java (hunk 10 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/gui/screens/settings/elements/CustomColorConfigOptionElement.java:32]: A square and fair defence! :)
